### PR TITLE
[Android] Implement embedding API: getHitTestResult()

### PR DESCRIPTION
**Pull request by chuan9**
Implement this API to get result when hit the screen

BUG=XWALK-6878

### CODE REVIEW

**Comment by chuan9** [comment]
@fujunwei @wuhengzhi Please have a review.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2016 Intel Corporation. All rights reserved.
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Add the copyright for upstream?

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package org.xwalk.core.internal;
+
+import java.io.BufferedWriter;
+
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Extra line.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import java.io.File;
+
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Extra line.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    public void setExtra(String extra) {
+        mExtra = extra;
+    }
+
+    /**
+     * Gets the type of the hit test result. See the XXX_TYPE constants
+     * defined in this class.
+     *
+     * @return the type of the hit test result
+     */
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Missing the api version.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    @XWalkAPI
+    public type getType() {
+        type a;
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Return `mType` directly?

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * Gets additional type-dependant information about the result. See
+     * {@link WebView#getHitTestResult()} for details. May either be null
+     * or contain extra information about this result.
+     *
+     * @return additional type-dependant information about the result
+     */
+    @XWalkAPI
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Missing the api version.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    public String getExtra() {
+        return mExtra;
+    }
+}
+
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Extra line.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java [hunk 8]
```diff
@@ -89,9 +89,22 @@
...
+      public int hitTestResultType;
+      public String hitTestResultExtraData;
+
+      // Used in requestFocusNodeHref(all three) and requestImageRef(only imageSrc)
+      public String href;
+      public String anchorText;
+      public String imgSrc;
+    }
+
+
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: Extra line.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java [hunk 9]
```diff
@@ -875,6 +902,16 @@ public void onGeolocationPermissionsHidePrompt() {
         mContentsClientBridge.onGeolocationPermissionsHidePrompt();
     }
 
+    @CalledByNative
+    private void updateHitTestData(
+            int type, String extra, String href, String anchorText, String imgSrc) {
+      mPossiblyStaleHitTestData.hitTestResultType = type;
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: Need four spaces.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java [hunk 10]
```diff
@@ -580,6 +584,17 @@ public void stopLoading() {
      * @return the url for current web page/app.
      * @since 1.0
      */
+    
+    @XWalkAPI
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java]: Above comment is for `getUrl`, so please adjust the order and add comment for `getHitTestResult`.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java [hunk 11]
```diff
@@ -580,6 +584,17 @@ public void stopLoading() {
...
+    public HitTestResultInternal getHitTestResult() {
+      if (mContent == null)  return null;
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java]: Indent, need four spaces.

File: runtime/browser/android/renderer_host/xwalk_render_view_host_ext.h [hunk 12]
```diff
@@ -40,9 +42,10 @@ class XWalkRenderViewHostExt : public content::WebContentsObserver,
   void ClearCache();
 
   // Do a hit test at the view port coordinates and asynchronously update
-  // |last_hit_test_data_|. |view_x| and |view_y| are in density independent
+  // test_data_|. Width and height in |touch_area| are in density independent
   // pixels used by WebKit::WebView.
-  void RequestNewHitTestDataAt(int view_x, int view_y);
+  void RequestNewHitTestDataAt(const gfx::PointF& touch_center,
+                                         const gfx::SizeF& touch_area);
```
Comment (wuhengzhi) [inline runtime/browser/android/renderer_host/xwalk_render_view_host_ext.h]: Indent.

File: runtime/browser/android/xwalk_content.cc [hunk 13]
```diff
@@ -263,6 +265,50 @@ void XWalkContent::Destroy(JNIEnv* env, jobject obj) {
...
+  ScopedJavaLocalRef<jstring> anchor_text;
+  if (data.anchor_text.length())
+    anchor_text = ConvertUTF16ToJavaString(env, data.anchor_text);
+
+  ScopedJavaLocalRef<jstring> img_src;
+  if (data.img_src.is_valid())
+    img_src = ConvertUTF8ToJavaString(env, data.img_src.spec());
+  Java_XWalkContent_updateHitTestData(env,
+                                      obj,
+				      data.type,
```
Comment (wuhengzhi) [inline runtime/browser/android/xwalk_content.cc]: Indent.

File: runtime/common/android/xwalk_render_view_messages.h [hunk 14]
```diff
@@ -11,6 +11,9 @@
 #include "ipc/ipc_message_macros.h"
 #include "ipc/ipc_platform_file.h"
 #include "third_party/skia/include/core/SkColor.h"
+#include "ui/gfx/geometry/point_f.h"
+#include "ui/gfx/geometry/size_f.h"
+
```
Comment (wuhengzhi) [inline runtime/common/android/xwalk_render_view_messages.h]: Extra line.

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 15]
```diff
@@ -29,6 +30,10 @@
 #include "third_party/skia/include/core/SkPicture.h"
 #include "xwalk/runtime/common/android/xwalk_hit_test_data.h"
 #include "xwalk/runtime/common/android/xwalk_render_view_messages.h"
+#include "url/url_canon.h"
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Alpha sort.

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 16]
```diff
@@ -29,6 +30,10 @@
...
+#include "url/url_constants.h"
+#include "url/url_util.h"
+
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Extra line.

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 17]
```diff
@@ -66,7 +71,12 @@ bool RemovePrefixAndAssignIfMatches(const base::StringPiece& prefix,
   const base::StringPiece spec(url.possibly_invalid_spec());
 
   if (spec.starts_with(prefix)) {
-    dest->assign(spec.begin() + prefix.length(), spec.end());
+	  url::RawCanonOutputW<1024> output;
+	  url::DecodeURLEscapeSequences(spec.data() + prefix.length(),
+									spec.length() - prefix.length(), &output);
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Indent.

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 18]
```diff
@@ -66,7 +71,12 @@ bool RemovePrefixAndAssignIfMatches(const base::StringPiece& prefix,
...
+	  *dest =
+		  base::UTF16ToUTF8(base::StringPiece16(output.data(), output.length()));
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Indent.

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 19]
```diff
@@ -66,7 +71,12 @@ bool RemovePrefixAndAssignIfMatches(const base::StringPiece& prefix,
...
+//    dest->assign(spec.begin() + prefix.length(), spec.end());
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Remove it?

File: runtime/renderer/android/xwalk_render_view_ext.cc [hunk 20]
```diff
@@ -207,13 +217,15 @@ void XWalkRenderViewExt::FocusedNodeChanged(const blink::WebNode& node) {
   Send(new XWalkViewHostMsg_UpdateHitTestData(routing_id(), data));
 }
 
-void XWalkRenderViewExt::OnDoHitTest(int view_x, int view_y) {
+void XWalkRenderViewExt::OnDoHitTest(const gfx::PointF& touch_center,
+                                          const gfx::SizeF& touch_area) {
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.cc]: Indent.

File: runtime/renderer/android/xwalk_render_view_ext.h [hunk 21]
```diff
@@ -37,7 +40,8 @@ class XWalkRenderViewExt : public content::RenderViewObserver {
 
   void OnDocumentHasImagesRequest(int id);
 
-  void OnDoHitTest(int view_x, int view_y);
+  void OnDoHitTest(const gfx::PointF& touch_center,
+                        const gfx::SizeF& touch_area);
```
Comment (wuhengzhi) [inline runtime/renderer/android/xwalk_render_view_ext.h]: Indent.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java [hunk 22]
```diff
@@ -0,0 +1,381 @@
+// Copyright (c) 2016 Intel Corporation. All rights reserved.
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java]: Add the copy for upstream?

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java [hunk 23]
```diff
@@ -0,0 +1,381 @@
...
+import android.graphics.BitmapFactory;
+import android.os.Message;
+import android.test.suitebuilder.annotation.LargeTest;
+import android.test.suitebuilder.annotation.SmallTest;
+import android.util.Log;
+
+import java.io.InputStream;
+import java.net.URL;
+import java.util.concurrent.Callable;
+import android.view.KeyEvent;
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java]: Alpha sort.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java [hunk 24]
```diff
@@ -0,0 +1,381 @@
...
+
+import org.chromium.base.test.util.Feature;
+import org.chromium.content.browser.ContentViewCore;
+import org.chromium.net.test.util.TestWebServer;
+
+import org.xwalk.core.internal.XWalkViewInternal;
+import org.xwalk.core.internal.xwview.test.util.CommonResources;
+import org.xwalk.core.internal.HitTestResultInternal;
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java]: Alpha sort.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java [hunk 25]
```diff
@@ -0,0 +1,381 @@
...
+        srcAnchorTypeRelativeUrlTestBody(true);
+    }
+
+    @SmallTest
+    @Feature({"XWalkView", "HitTest"})
+    public void testSrcAnchorTypeRelativeUrlByFocus() throws Throwable {
+        srcAnchorTypeRelativeUrlTestBody(false);
+    }
+
+ private void srcEmailTypeTestBody(boolean byTouch) throws Throwable {
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/HitTestTest.java]: Indent.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java [hunk 26]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2016 Intel Corporation. All rights reserved.
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java]: Add the copyright for upstream.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+                view.onTouchEvent(MotionEvent.obtain(
+                        eventTime, eventTime, MotionEvent.ACTION_DOWN,
+                        x, y, 0));
+                view.onTouchEvent(MotionEvent.obtain(
+                        eventTime, eventTime, MotionEvent.ACTION_UP,
+                        x, y, 0));
+            }
+        });
+    }
+
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java]: Extra line.

File: test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+}
+
```
Comment (wuhengzhi) [inline test/android/core_internal/javatests/src/org/xwalk/core/internal/xwview/test/XWalkViewInternalTestTouchUtils.java]: ditto.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 4 above)
Comment (chuan9) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: This is not a XWalkAPI, which version do you mean?

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 1 above)
Comment (chuan9) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: @wuhengzhi The copyright is same as others files just 2016 is diffirent, Is there any problem?

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 5 above)
Comment (chuan9) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: yes, return it directly

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 1 above)
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: If the code was ported from upstream, we should add copyright of upstream.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 4 above)
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: `@since 7.0`

**Comment by chuan9** [comment]
@wuhengzhi @fujunwei Done, thanks for your review.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+     * Gets the type of the hit test result. See the XXX_TYPE constants
+     * defined in this class.
+     *
+     * @return the type of the hit test result
+     * @since 7.0
+     */
+    @XWalkAPI
+    public type getType() {
+        type a;
+        switch (mType){
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: Return `mType` ?

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java [hunk 30]
```diff
@@ -875,6 +901,16 @@ public void onGeolocationPermissionsHidePrompt() {
         mContentsClientBridge.onGeolocationPermissionsHidePrompt();
     }
 
+    @CalledByNative
+    private void updateHitTestData(
+        int type, String extra, String href, String anchorText, String imgSrc) {
+      mPossiblyStaleHitTestData.hitTestResultType = type;
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkContent.java]: Indent.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java [hunk 31]
```diff
@@ -588,6 +592,20 @@ public String getUrl() {
     }
 
     /**
+    * @return the result of the hit place in the current page.
```
Comment (wuhengzhi) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewInternal.java]: Missing the explanation for this function.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+// found in the LICENSE file.
+
+package org.xwalk.core.internal;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.util.Map;
+
+@XWalkAPI(createInternally = true)
+public class HitTestResultInternal {
```
Comment (fujunwei) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: @lincsoon, Do the name need to add prefix `xwalk`.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 29 above)
Comment (chuan9) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: @wuhengzhi no, return a, mType is an integer type which is used by internal. we should convert integer to type here.

File: runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java (hunk 32 above)
Comment (chuan9) [inline runtime/android/core_internal/src/org/xwalk/core/internal/HitTestResultInternal.java]: @fujunwei Should we change this name to XWalkHitTestResultInternal? if so, developper use "XWalkHitTestResult ** = ((XWalkView) v).getHitTestResult();", while webview use case is "
HitTestResult ** = ((XWalkView) v).getHitTestResult();", while webview use case is "
WebView.HitTestResult result = ((WebView) v).getHitTestResult();"

**Pull request merged**